ic-agent/src/agent/mod.rs

```diff
@@ -890,16 +914,17 @@ impl Agent {
         &self,
         delegation: &Option<Delegation>,
         effective_canister_id: Principal,
+        root_key: Vec<u8>,
     ) -> Result<Vec<u8>, AgentError> {
         match delegation {
-            None => Ok(self.read_root_key()),
+            None => Ok(root_key),
```

Comment:
What is the reason for passing `root_key` as a parameter here, rather than getting it from `self`?

Reply:
verify_cert & check_delegation would become async and you have a recursive future. I wasn't able to unroll it into an iterator unfortunately, so I went with this

Reply:
verify_cert & check_delegation would become async and you have a recursive future. I wasn't able to unroll it into an iterator unfortunately, so I went with this